Fix blob_get() would timeout. Issue #2070

cristi-zz opened the pull request:
Tentative fixing. The node is primed with some fixed peers before triggering the blob download.

Please check comments on issue #2070

tzarebczan commented:
@cristi-zz nice to see your face here - we've come full circle since, huh! 

Thanks for the PR and congrats on the first one, we'll get this reviewed soon. 

Can we show you some [appreciation](https://lbry.com/faq/appreciation)?

cristi-zz commented:
@tzarebczan thank you for the appreciation!

This bug was on my mind for so long! However the fix is ugly (duplicate code, etc), some guidance is needed to proper fix this. My guess is that peers must be added to the Node object before interacting with Kademlia, and this "priming" should happen at the startup of the library and not be rebooted each time an API call finishes. Now, the peers are added in some API handling routines (download a stream with `get` ) but not in others (like `blob_get`). Don't know what is the design intention so I can't dig further.

cristi-zz commented:
Ready for review. All tests (int+unit) are green on local machine.

lyoshenka commented:
@jackrobison please review this issue. it's been open for over a month

jackrobison commented on lbry/dht/node.py:
```diff
@@ -258,3 +258,13 @@ def accumulate_peers(self, search_queue: asyncio.Queue,
+                              for address, port in peer_address_list]
+        peer_queue.put_nowait(kademlia_peer_list)
+    except asyncio.CancelledError:
+        pass
```
why is the cancelled error trapped? usually trapping this error causes unexpected behavior.

cristi-zz replied:
answer in the comments of this PR

jackrobison reviewed (changes requested):
I'm not understanding what the desired change in behavior here is. It looks like the `call_later` is refactored to be a `asyncio.Task` with a `asyncio.sleep` instead of the call_later delay - which should behave essentially the same.

Can you explain a bit what the problem is / what the desired behavior looks like?

cristi-zz commented:
Hi! Thank you for looking into this! This bug, is bugging me for a while, now.

Overall intention:

Make RPC call ``blob_get`` connect to some peers and get the actual blobs. Imho there were some refactoring on the way the DHT operations are handled and this particular method was left out. Fixing it directly would result in duplicate code (true, 2-3 lines only but still, a new source of grief). So I tried to extract the common behavior so future changes in the semantics of DHT will not break these calls. 

Mechanics:
In the original code the ``_delayed_add_fixed_peers`` was a regular method that was called after ``fixed_peers_delay``. Also, the handle returned by ``call_later`` is used to cancel this task [why the delay? why the need to cancel? Not important right now, maybe]. I tried to reproduce the exact behavior , constrained by the facts:
(1) the _delayed_add_fixed_peers must call an async code (so it must be also made async. The ``call_later`` does not take a coroutine) and (2) not being an expert on asyncio. 

Reading [loop.call_later](https://docs.python.org/3.7/library/asyncio-eventloop.html#asyncio.loop.call_later) docs I assume that if the task is cancelled before the delay, no exception is thrown. I THINK I tried to mimic the same behavior in ``prime_node_with_peer_addresses`` by catching only the Cancelled exception. Looking back at the code, maybe is the wrong place to catch it? Where this exception would propagate? To ``add_fixed_peers``? Is there a more pythonic way to add peers after a delay while being able to cancel the whole operation? 

Suggestions are more than welcome wrt to this issue!

cristi-zz commented on lbry/stream/downloader.py:
```diff
@@ -48,26 +48,18 @@ async def cached_read_blob(blob_info: 'BlobInfo') -> bytes:
             self.fixed_peers_delay = self.config.fixed_peer_delay
-
-        self.fixed_peers_handle = self.loop.call_later(self.fixed_peers_delay, _delayed_add_fixed_peers)
+        self.fixed_peers_handle = self.loop.call_later(self.fixed_peers_delay, asyncio.create_task, _add_fixed_peers())
```
imho it would be cleaner to have ``sleep`` followed by ``_add_peers`` and no cancel handle. But maybe I am missing some important info and the cancelling handle should stay in place. Anyway, it is more asyncio-ish than before.

cristi-zz commented:
Hi! @jackrobison !

Refactored the offending line [see comments on commit] and removed the try/catching the Cancel exception. Also don't know if you get notified at comments or only when "resolve conversation" is pressed. So sorry if I spammed you with tags.

Cheers!

jackrobison commented on tests/unit/blob_exchange/test_transfer_blob.py:
```diff
@@ -326,3 +329,33 @@ async def sendfile(writer):
+                    daemon_mock_dht.return_value = mock_node
+                    daemon_mock_blob_manager.return_value = self.client_blob_manager
+                    result = await daemon.jsonrpc_blob_get(blob_hash)
+                    self.assertIsNotNone(result)
```
this could return an error and still pass - should check that the result is expected

cristi-zz replied:
Checked the downloaded blob against the "original" one, injected in the server. HOWEVER, the original bug is timeout. Didn't look into the codebase to see how it is done here, and saw that unittest does not handle the timeout natively [sth like timeout_decorator]. CI will handle timeouts anyway so this test will fail on CI (if ofc the peers will not be added correctly). Not local maybe. Want to add a depedency? Hack/simulate the timeout check using some async? (make a code that sleep for x seconds and assertFalse. Kill the future after blob_get is done). Or good enough for now?

Also, saw some coverage reductions, but I guess they are from incoming code.

cristi-zz commented:
Hm, it worked on my github instance, Maybe retrigger the CI pipeline?

https://github.com/cristi-zz/lbry-sdk/actions/runs/1420543979
